### ISSUE TITLE
tf.estimator.EstimatorSpec doesn't have evaluation_hooks parameter

### ISSUE THREAD

**Issue by hanfeisun** [tensorflow/tensorflow#14042]
I planned to visualize the evaluation result in tensorboard. Therefore I need to create the evaluation_hook using `tf.train.SummarySaverHook` in model_fn, and pass it into the [EstimatorSpec](https://www.tensorflow.org/api_docs/python/tf/estimator/EstimatorSpec). 

However, [EstimatorSpec](https://www.tensorflow.org/api_docs/python/tf/estimator/EstimatorSpec) doesn't accept `evaluation_hooks` for now. It only has `training_hooks`. Will `evaluation_hooks` be added in future versions?

**Comment by hanfeisun** [comment]
Though [Estimator.evaluate](https://www.tensorflow.org/api_docs/python/tf/estimator/Estimator#evaluate) has `hooks` parameter. It is outside of the model_fn context and can't get the `summary_op` conveniently. It will be great if EstimatorSpec could have `evaluation_hooks` parameter.

**Comment by skye** [comment]
@martinwicke FYI

**Comment by martinwicke** [comment]
@ispirmustafa, WDYT

**Comment by mtngld** [comment]
During training mode of an Estimator the summaries are added automatically (no need to use hooks). Why isn't it the same behavior for evaluation mode?

Related StackOverflow question: https://stackoverflow.com/questions/43782767/how-can-i-use-tensorboard-with-tf-estimator-estimator

**Comment by ispirmustafa** [comment]
It makes sense for consistency with training.
On the other hand I haven't see a valid eval hook request. @hanfeisun how are you planning to use SummaryHook in evaluation? There are more then one eval-run in evaluation of one global-step.

**Comment by chahld** [comment]
I've created a kludgey solution, by modifying the SummarySaverHook code and getting the summary op via tf.get_collections(tf.GraphKeys.SUMMARY_OP).  In my use case I want only one summary saved (to show example images and neuron response images. But you can probably modify this to do whatever you want. 

You pass this into the estimator.evaluate like this:
`
    hooks = [EvalSummarySaverHook(output_dir=eval_model_dir)]
    eval_results = eval_estimator.evaluate(input_fn=eval_input_func, hooks=hooks)
`

Here is the code:
class EvalSummarySaverHook(tf.train.SessionRunHook):
    """Saves summaries during eval loop."""

    def __init__(self,
                 output_dir=None,
                 stop_after=1):
        """Initializes a special `SummarySaverHook` to run during evaluations

        Args:
          output_dir: `string`, the directory to save the summaries to.
        """
        self._summary_op = None
        self._output_dir = output_dir
        self._global_step_tensor = None
        self._saves_remaining = stop_after

    def begin(self):
        self._global_step_tensor = tf.train.get_or_create_global_step()
        if self._global_step_tensor is None:
            raise RuntimeError(
                "Global step should be created to use SummarySaverHook.")

    def before_run(self, run_context):  # pylint: disable=unused-argument
        requests = {"global_step": self._global_step_tensor}
        if self._saves_remaining and self._get_summary_op() is not None:
            requests["summary"] = self._get_summary_op()

        return tf.train.SessionRunArgs(requests)

    def after_run(self, run_context, run_values):
        _ = run_context
        if "summary" in run_values.results:
            print('Saving eval summaries')
            global_step = run_values.results["global_step"]
            summary_writer = tf.summary.FileWriterCache.get(self._output_dir)

            for summary in run_values.results["summary"]:
                summary_writer.add_summary(summary, global_step)

            self._saves_remaining -= 1

    def end(self, session=None):
        _ = session
        summary_writer = tf.summary.FileWriterCache.get(self._output_dir)
        summary_writer.flush()

    def _get_summary_op(self):
        """Fetches the summary op from collections
        """

        if self._summary_op is None:
            self._summary_op = tf.get_collection(tf.GraphKeys.SUMMARY_OP)

        return self._summary_op
`

**Comment by huw** [comment]
The API lists `evaluation_hooks` as a [valid parameter](https://www.tensorflow.org/api_docs/python/tf/estimator/EstimatorSpec) for `EstimatorSpec` — does this not work?

**Comment by hanfeisun** [comment]
@mtngld I need the summary hook because I want to config `save_steps` and `output_dir` parameter using SummarySaverHook.  My codes looks like this:

```
    if mode == tf.estimator.ModeKeys.TRAIN:
        tf.summary.image("original", features['source'])
        tf.summary.image("target", features['target'])
        tf.summary.image("transfer_sigmoid", generated)
    else:
        # Evaluation
        tf.summary.image("eval_original", features['source'], max_outputs=10)
        tf.summary.image("eval_target", features['target'], max_outputs=10)
        tf.summary.image("eval_transfer_sigmoid", generated, max_outputs=10)
    summary_hook = tf.train.SummarySaverHook(
        save_steps=10,
        output_dir="./board",
        summary_op=tf.summary.merge_all())
    return tf.estimator.EstimatorSpec(
        mode=mode,
        loss=loss,
        train_op=tf.group(d_train, g_train),
        eval_metric_ops=None,
        training_hooks=[summary_hook]
    )
```

As I remember, during the evaluation, the image summary is not shown in tensorboard somehow..

**Comment by hanfeisun** [comment]
@huw That looks great! I think the API is newly added as there is no `evaluation_hooks` yet in "[tensorflow 1.3](evaluation_hooks)"

**Comment by bodokaiser** [comment]
I would love to get this working with the higher level APIs [`tf.estimator.train_and_eval`][1] and [`Estimator.evaluate`][2]

1: https://www.tensorflow.org/api_docs/python/tf/estimator/train_and_evaluate
2: https://www.tensorflow.org/api_docs/python/tf/estimator/Estimator#evaluate

**Comment by bhack** [comment]
@martinwicke Have you assigned this after was already closed? Is there a plan to cover @bodokaiser  request?

**Comment by martinwicke** [comment]
I don't think there's anything else to do:

- `train_and_evaluate` uses EvalSpec, which has a `hooks` field.
- `Estimator.evaluate` has `hooks` argument.

**Comment by bodokaiser** [comment]
I see this goes on in #15332 .
